Implement config backup cli

daniel-illi opened the pull request:
This is a work in progress for an implementation of #1382.

daniel-illi commented:
I have extracted the config backup code from [`src/rockstor/storageadmin/views/config_backup.py`](https://github.com/rockstor/rockstor-core/pull/1576/files#diff-c759675f7d5bdaa63b9c5b220f2521ce) to its own file [`src/rockstor/system/config_backup.py`](https://github.com/rockstor/rockstor-core/pull/1576/files#diff-8ebc72135a50100d430dbc652be8d474) and have verified that the config backup in the admin gui still works.

I have implemented a basic [cli script](https://github.com/rockstor/rockstor-core/pull/1576/commits/9e07cfeb899926824d8e824aafc1e6875792f96b) to trigger the config backup, but when I run it from the shell, I get a stacktrace:

```
# bin/backup-config 
Traceback (most recent call last):
  File "bin/backup-config", line 39, in <module>
    sys.exit(scripts.config_backup.main())
  File "/opt/build/src/rockstor/scripts/config_backup.py", line 24, in main
    result = backup_config()
  File "/opt/build/src/rockstor/system/config_backup.py", line 49, in backup_config
    call_command('dumpdata', *model_list, stdout=dfo)
  File "/opt/build/eggs/Django-1.8.16-py2.7.egg/django/core/management/__init__.py", line 92, in call_command
    app_name = get_commands()[name]
  File "/opt/build/eggs/Django-1.8.16-py2.7.egg/django/utils/lru_cache.py", line 101, in wrapper
    result = user_function(*args, **kwds)
  File "/opt/build/eggs/Django-1.8.16-py2.7.egg/django/core/management/__init__.py", line 72, in get_commands
    for app_config in reversed(list(apps.get_app_configs())):
  File "/opt/build/eggs/Django-1.8.16-py2.7.egg/django/apps/registry.py", line 137, in get_app_configs
    self.check_apps_ready()
  File "/opt/build/eggs/Django-1.8.16-py2.7.egg/django/apps/registry.py", line 124, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

It seems I didn't initialize the application correctly.
@schakrava or anyone else, can you help me with this?

phillxnet commented:
@daniel-illi I'm afraid this one is more in the domains of @MFlyer or @schakrava but it's great you've made a start on it. Should be a really useful feature once done.

schakrava commented:
Can one of the admins verify this patch?

MFlyer commented:
@schakrava : 

Patched over dev env, here my output:

```
Traceback (most recent call last):
  File "./backup-config", line 39, in <module>
    sys.exit(scripts.config_backup.main())
  File "/opt/build/src/rockstor/scripts/config_backup.py", line 25, in main
    print('config exported to: "%s"' % result.filename)
AttributeError: 'ReturnDict' object has no attribute 'filename'
```

there's only a wrong dict

MFlyer commented on src/rockstor/scripts/config_backup.py:
```diff
@@ -0,0 +1,29 @@
+
+def main():
+    result = backup_config()
+    print('config exported to: "%s"' % result.filename)
```
@daniel-illi - @schakrava 

Check my comment on pr, move `result.filename` to `result['filename']`

Mirko

MFlyer replied:
Hi @daniel-illi can you rebase & patch this with that change?

M.

MFlyer commented:
Backup nicely exists on WebUI too, @daniel-illi have you checked system restore?
M.

daniel-illi commented:
@MFlyer thanks for the feedback.
Sure I will rebase and apply your proposed change.
I'm away for the weekend, will do it next week.

schakrava commented:
Can one of the admins verify this patch?

daniel-illi commented:
rebase done and patch applied.
will have to test next week...

MFlyer commented:
@schakrava on my side we can go with this
Mirko

MFlyer commented on src/rockstor/storageadmin/views/config_backup.py:
```diff
@@ -184,35 +185,9 @@ def get_queryset(self, *args, **kwargs):
-            cbo = ConfigBackup(filename=gz_name, md5sum=md5sum, size=size)
-            cbo.save()
-            return Response(ConfigBackupSerializer(cbo).data)
+            return Response(backup_config(self.cb_dir, self._md5sum))
```
@daniel-illi 
this `return Response(backup_config(self.cb_dir, self._md5sum))` should be
`return Response(backup_config())`

Your backup_config doesn't have args, found it while having some backup tests over another issue 😉

MFlyer commented:
@schakrava mods required, pls wait merging (review on code)

MFlyer commented:
Adds:
failing over WebUI restore,
https://github.com/daniel-illi/rockstor-core/blob/50c079b237bedcef24583c1ea4e5671e7cdb4371/src/rockstor/storageadmin/views/config_backup.py#L210 must be `self._validate_input(backup_id, request)` - not @daniel-illi , suspicion about previous mods, checking
and getting `ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/sm/services/d`

MFlyer commented on src/rockstor/system/config_backup.py:
```diff
@@ -0,0 +1,68 @@
+    return ConfigBackupSerializer(cbo).data
+
+
+class ConfigBackupMixin(object):
```
@daniel-illi while checking flake8 got `ConfigBackupMixin(object)` : here you have it both in system and Django view

My coding style suggestion : move md5sum to an upper level (ex. system osy.py or util.py, so all contributors can get it all arount Rockstor code 😉 ), import it in Django view and here (so doing we avoid calling that _md5sum private function, bad practice), avoid ConfigBackupSerializer (your cli needs online filename, get it directly! :) )

@schakrava @phillxnet ? suggestions?

MFlyer commented:
Exceptions: current config_backup runs with rest from cli, moved to APIWrapper and all got solved.
#1585 patching requires some testing to correctly grant both samba global config saving and config backup/restore, working on it

MFlyer commented:
@schakrava  #1639 fixing #1585 on its way, ready to be merged without this PR

Anyway this PR needs suggested mods + APIWrapper (config restore not working with rest util), waiting @daniel-illi reply to patch APIWrapper over his fork

daniel-illi commented:
@MFlyer thanks for your feedback.
I'll try to integrate ApiWrapper as requested.

MFlyer commented:
Good morning @daniel-illi , while testing your PR had code to convert rest cli to APIWrapper, please tell us if you need help :)

Note about APIWrapper: check over smart_manager/data_collector for APIWrapper import and use (pls. note it's rest cli "younger brother" and url have a different pattern : ex. samba is just 'samba', no slashes and no baseurl required)

Mirko

MFlyer commented:
Hi @daniel-illi , any news on this?

Mirko

daniel-illi commented:
Hi @MFlyer, sorry I didn't find time yet besides work and family responsibilities.
I'll have some time next weekend and I'm planning to finish this feature.

daniel-illi commented:
Today I rebased on the current master branch and fixed some minor error.
Everything seems to work both from the command line and with the web gui.

@MFlyer while working on this a bit today, I got confused about where the ApiWrapper is needed. Can you please point me to which calls need to be migrated? Thanks!

MFlyer commented:
Hi @daniel-illi , happy to see you back on this :)

I need to recap - meanwhile had code over other issues -, but I'll be back asap (probably by tomorrow 😉), it's a promise.
Mirko

daniel-illi commented:
@MFlyer that would be great but take your time, it will be a while until I can work on this again.

daniel-illi commented:
Had another minute to work on this pr:
* Rebased on current upstream master
* Script now prints the full path to the generated config backup file
* Tested generating config backup from command line and from web gui
* Tested restoring config from generated backup file

From my point of view the feature is ready for merge.

@MFlyer can you check it please?

MFlyer commented:
Hi @daniel-illi , thanks for your update, will check it asap 😃 

Mirko

phillxnet commented:
@daniel-illi Are you still of the opinion that this is ready for merge? I know @MFlyer was instrumental in some of this, early review wise. Only asking as it would be great to get this into the current testing channel updates. You state that a restore test was carried out, did this successfully restore, on the most recent fresh iso install, at least one and ideally 2 -3 elements of each of the current claims re restore coverage in:
http://rockstor.com/docs/config_backup.html#config-restore

I'm personally chuffed by this feature / enhancement and quite looking forward to getting it in place. It's going to be a nice peace of mind element to config backup and restore. Especially given the fixes included here also.

I can update the docs once it's merged to reflect the command line capability; can you include a quick pointer on the intended command line use within this pr?

Cheers.

phillxnet commented:
I have linked to this pr from issue #1057 to identify this pr as a dependency of the linked issue. I.e. this pr should be merged before the referenced issue is addressed.

daniel-illi commented:
@phillxnet when I last worked on this one month ago, it looked good to me. I didn't rebase it since then and don't really have any time at the moment as any day now I will become a father for the second time :)

I tested backup and restore on a fresh install with the (then) latest iso and the setup as describe in http://rockstor.com/docs/contribute.html

If I remember correctly, I tested restoring the configs of the user/groups, samba shares, afp shares, and nfs exports.
I did not test extensively, and not at all the services configuration. But I doubt there will be any problem (assuming it worked before) as I did not change the code which serializes the configurations.

Usage so far is really simple, you just call the script and it prints the path to the exported config file:
```
# /opt/rockstor/bin/backup-config 
config exported to: "/opt/rockstor/static/config-backups/backup-2017-06-16-230709.json.gz"
```

Ultimately I'd like the default backup location to be configurable from the web interface so I could store the backups in a separate share. And it might be useful if I could specify a path as an argument for the backup-config script. But these features can wait for another time..

MFlyer commented:
@daniel-illi - @phillxnet : had tests on this long time ago and it was working, will have a new test today

M.

phillxnet commented:
@daniel-illi Thanks for the issue update and congratulations on the continuing population / propagation efforts. Agreed the additional niceties can wait as this is a major enhancement as is. Looks like @MFlyer is up for a re-look so thanks to @MFlyer also.

MFlyer commented:
Sorry for the delay, KO on my side @phillxnet & @schakrava : 

```
root@rockstone:bin# tail -f /opt/build/var/log/rockstor.log 

[18/Jun/2017 05:36:52] DEBUG [storageadmin.views.config_backup:215] restore starting...
[18/Jun/2017 05:36:52] DEBUG [storageadmin.views.config_backup:217] restore submitted...
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:55] Started restoring users and groups.
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/groups. payload: {u'admin': True, u'groupname': u'rocky', u'gid': 1001}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/users. payload: {u'username': u'rocky', u'public_key': None, u'shell': u'/bin/bash', u'group': 1, u'uid': 1001, u'admin': True, u'homedir': u'/home/rocky', u'gid': 1001, u'user': 1, 'password': 'rockstor', u'email': None, u'smb_shares': []}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:71] Finished restoring users and groups.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:75] Started restoring Samba exports.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:83] Finished restoring Samba exports.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:102] Started restoring NFS exports.
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/adv-nfs-exports. payload: {'entries': []}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:122] Finished restoring NFS exports.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:87] Started restoring AFP exports.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:98] Finished restoring AFP exports.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:126] Started restoring services.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:135] services = {u'rockstor': {'config': {u'network_interface': u'ens3', u'listener_port': u'443'}}, u'smb': {'config': {}}, u'shellinaboxd': {'config': {u'detach': False, u'css': u'white-on-black', u'shelltype': u'LOGIN'}}}
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/sm/services/rockstor/config. payload: {'config': {u'network_interface': u'ens3', u'listener_port': u'443'}}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/sm/services/smb/config. payload: {'config': {}}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] ERROR [storageadmin.views.config_backup:51] Exception occured while creating resource: https://localhost/api/sm/services/shellinaboxd/config. payload: {'config': {u'detach': False, u'css': u'white-on-black', u'shelltype': u'LOGIN'}}. exception: HTTPSConnectionPool(host='localhost', port=443): Max retries exceeded with url: /o/token/ (Caused by <class 'socket.error'>: [Errno 111] Connection refused). Moving on.
[18/Jun/2017 05:36:53] DEBUG [storageadmin.views.config_backup:138] Finished restoring services. 
```

@daniel-illi can you check it? that was a config restore via web ui
M.

daniel-illi commented:
@MFlyer I won't have time to look into this for the next approx. 4 weeks as I'll be on paternity leave any day now. Probably even today :smiley:

schakrava commented:
LGTM. Thank you @daniel-illi. Wishing you a wonderful next 4 weeks!

Thanks @MFlyer and @phillxnet for your reviews.
